File: app/src/main/java/com/angogasapps/myfamily/ui/screens/buy_list/BuyListFragment.java
package com.angogasapps.myfamily.ui.screens.buy_list;

import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.angogasapps.myfamily.databinding.FragmentBuyListBinding;
import com.angogasapps.myfamily.models.buy_list.BuyList;
import com.angogasapps.myfamily.models.buy_list.BuyListEvent;
import com.angogasapps.myfamily.ui.screens.buy_list.adapters.ProductsAdapter;
import com.angogasapps.myfamily.ui.screens.buy_list.dialogs.BuyListProductCreatorDialog;

import io.reactivex.disposables.Disposable;


public class BuyListFragment extends Fragment {
    private FragmentBuyListBinding binding;
    private BuyList buyList;
    private ProductsAdapter adapter;
    private LinearLayoutManager layoutManager;
    private Disposable observer;


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentBuyListBinding.inflate(inflater, container, false);
        initOnClickListeners();
        initObserver();
        initRecyclerView();

//        binding.buyListName.setText(buyList.getName());

        getActivity().setTitle(buyList.getName());

        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (observer != null)
            if (!observer.isDisposed())
                observer.dispose();
    }

    private void initOnClickListeners() {
        binding.floatingBtn.setOnClickListener(v -> {
            (new BuyListProductCreatorDialog(getContext(), buyList.getId())).show();
        });
    }

    private void initObserver() {
        observer = BuyListManager.getInstance().subject.subscribe(event -> {
            if (event.getEvent().equals(BuyListEvent.EBuyListEvents.buyListRemoved)){
                if (event.getBuyListId().equals(this.buyList.getId())){
                    onRemoveThisBuyList();
                    return;
                }
            }
            if (!event.getBuyListId().equals(this.buyList.getId())) {
                return;
            }
            adapter.update(event);
        });
    }

    private void initRecyclerView() {
        adapter = new ProductsAdapter(getContext(), buyList);
        layoutManager = new LinearLayoutManager(getContext());
        binding.recycleView.setLayoutManager(layoutManager);
        binding.recycleView.setAdapter(adapter);
    }

    private void onRemoveThisBuyList() {
        BuyListActivity.getIgoToListOfBuyListsFragment().go();
    }

    public BuyList getBuyList() {
        return buyList;
    }

    public void setBuyList(BuyList buyList) {
        this.buyList = buyList;
    }
}

File: app/src/main/java/com/angogasapps/myfamily/ui/screens/buy_list/BuyListFragment.kt
package com.angogasapps.myfamily.ui.screens.buy_list


import com.angogasapps.myfamily.ui.screens.buy_list.BuyListActivity.Companion.igoToListOfBuyListsFragment
import com.angogasapps.myfamily.models.buy_list.BuyList
import com.angogasapps.myfamily.ui.screens.buy_list.adapters.ProductsAdapter
import androidx.recyclerview.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.view.View
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.angogasapps.myfamily.app.appComponent
import com.angogasapps.myfamily.databinding.FragmentBuyListBinding
import com.angogasapps.myfamily.ui.screens.buy_list.dialogs.BuyListProductCreatorDialog
import com.angogasapps.myfamily.ui.screens.buy_list.BuyListManager
import com.angogasapps.myfamily.models.buy_list.BuyListEvent
import com.angogasapps.myfamily.network.repositories.BuyListRepository
import com.angogasapps.myfamily.ui.screens.buy_list.BuyListActivity
import io.reactivex.disposables.Disposable
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import javax.inject.Inject

class BuyListFragment : Fragment() {
    private lateinit var binding: FragmentBuyListBinding
    lateinit var buyList: BuyList
    private lateinit var adapter: ProductsAdapter
    private lateinit var layoutManager: LinearLayoutManager

    @Inject
    lateinit var buyListRepository: BuyListRepository

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        appComponent.inject(this)
        binding = FragmentBuyListBinding.inflate(inflater, container, false)
        initOnClickListeners()
        initObserver()
        initRecyclerView()
        requireActivity().title = buyList.name
        return binding.root
    }


    private fun initOnClickListeners() {
        binding.floatingBtn.setOnClickListener { v: View? ->
            BuyListProductCreatorDialog(
                requireContext(), buyList.id
            ).show()
        }
    }

    private fun initObserver() {
        lifecycleScope.launch {
            buyListRepository.listener.collect { event: BuyListEvent ->
                if (event.event == BuyListEvent.EBuyListEvents.buyListRemoved) {
                    if (event.buyListId == buyList.id) {
                        onRemoveThisBuyList()
                        return@collect
                    }
                }
                if (event.buyListId != buyList.id) {
                    return@collect
                }
                adapter.update(event)
            }
        }
    }

    private fun initRecyclerView() {
        adapter = ProductsAdapter(requireContext(), buyList)
        layoutManager = LinearLayoutManager(context)
        binding.recycleView.layoutManager = layoutManager
        binding.recycleView.adapter = adapter
    }

    private fun onRemoveThisBuyList() {
        igoToListOfBuyListsFragment.invoke()
    }
}